cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll

```diff
@@ -133,17 +133,35 @@ private module PartialDefinitions {
   }
 
   /**
-   * A partial definition that's a definition by reference (in the sense of the
-   * `definitionByReference` predicate).
+   * A partial definition that's a definition by reference.
    */
   class DefinitionByReference extends PartialDefinition, TReferenceArgument {
-    VariableAccess va;
-
-    DefinitionByReference() { definitionByReference(va, definedExpr) }
-
-    VariableAccess getVariableAccess() { result = va }
+    Expr getArgument() { this = TReferenceArgument(result, _) }
+  }
 
-    override predicate partiallyDefines(Variable v) { va = v.getAnAccess() }
+  private predicate referenceArgument(Expr argument, VariableAccess va) {
+    argument = any(Call c).getAnArgument() and
+    exists(Type argumentType |
+      argumentType = argument.getFullyConverted().getType().stripTopLevelSpecifiers()
+    |
+      argumentType instanceof ReferenceType and
+      not argumentType.(ReferenceType).getBaseType().isConst() and
```

Comment:
What about rvalue references to non-const types?

Reply:
I believe those are ruled out with this predicate. The argument is restricted to be a `VariableAccess`, and a `VariableAccess` is never an rvalue unless you send it through `std::move` (or cast it yourself, but that should be very rare).